category/vm/compiler/ir/x86/emitter.cpp

```diff
@@ -4373,62 +4397,254 @@ namespace monad::vm::compiler::native
-                as_.mov(byte_ix.r32(), gpq[0].r32());
-            }
-            as_.cmovnz(byte_ix.r32(), bound_mem);
-            as_.neg(byte_ix);
-            stack_mem = x86::qword_ptr(x86::rsp, byte_ix, 0, byte_off);
-        }
-        else if (ix->avx_reg()) {
-            as_.vmovd(x86::eax, avx_reg_to_xmm(*ix->avx_reg()));
-            as_.cmovnz(x86::eax, bound_mem);
-            as_.neg(x86::rax);
-            stack_mem = x86::qword_ptr(x86::rsp, x86::rax, 0, byte_off);
+        stack_.push(std::move(dst));
+    }
+
+    void Emitter::signextend_general_reg_by_bounded_rax(StackElemRef src)
+    {
+        MONAD_VM_DEBUG_ASSERT(src->general_reg().has_value());
+
+        GeneralRegReserv const src_reserv{src};
+        auto [dst, dst_reserv] = alloc_general_reg();
+
+        x86::Gpq shift;
+        std::variant<x86::Gpd, x86::Mem> shift63;
+        if (stack_.has_free_general_reg()) {
+            auto [e, reserv] = alloc_general_reg();
+            auto const &gpq = general_reg_to_gpq256(*e->general_reg());
+            shift = gpq[0];
+            shift63 = gpq[1].r32();
+            as_.mov(gpq[1].r32(), 63);
         }
         else {
-            MONAD_VM_DEBUG_ASSERT(ix->stack_offset().has_value());
-            auto mem = stack_offset_to_mem(*ix->stack_offset());
-            as_.mov(x86::eax, mem);
-            as_.cmovnz(x86::eax, bound_mem);
-            as_.neg(x86::rax);
-            stack_mem = x86::qword_ptr(x86::rsp, x86::rax, 0, byte_off);
+            as_.push(reg_context);
+            shift = reg_context;
+            shift63 = rodata_.add4(63);
+        }
+
+        auto cmovb_shift = [&, this] {
+            if (std::holds_alternative<x86::Gpd>(shift63)) {
+                as_.cmovb(shift.r32(), std::get<x86::Gpd>(shift63));
+            }
+            else {
+                as_.cmovb(shift.r32(), std::get<x86::Mem>(shift63));
+            }
+        };
+
+        auto const &src_gpq = general_reg_to_gpq256(*src->general_reg());
+        auto const &dst_gpq = general_reg_to_gpq256(*dst->general_reg());
+
+        // It is a pre-condition that rax is bounded, rax <= 31.
+
+        as_.lea(shift.r32(),
+            x86::byte_ptr(std::bit_cast<uint32_t>(int32_t{-56}), x86::eax, 3));
+        // shift.r32() = -56 + eax * 8
+        as_.neg(shift.r32());
+        // shift.r32() = 56 - eax * 8
+        as_.and_(shift.r32(), 63);
+        // shift.r32()
+        //   = (56 - eax * 8) % 64
+        //   = (56 % 64 - (eax * 8) % 64) % 64
+        //   = (56 - (eax * 8) % 64) % 64
+        //   = 56 - (eax * 8) % 64, because (eax * 8) % 64 <= 56
+        //   = 56 - (eax % 8) * 8
+        //   where the last equality follows from
+        //     8 * (eax / 8) + (eax % 8) = eax iff
+        //     64 * (eax / 8) + 8 * (eax % 8) = 8 * eax iff
+        //     64 * ((8 * eax) / 64) + 8 * (eax % 8) = 8 * eax
+        //   which implies that 8 * (eax % 8) = (8 * eax) % 64,
+        //   because 8 * (eax % 8) < 64.
```

Comment:
Comments with example input has been added to the general reg signextend algo. I have additionally found a nice optimization here.